File: backend/processPDFs.js
const fs = require('fs');
const path = require('path');
const pdf = require('pdf-parse');
const mongoose = require('mongoose');
const Specialty = require('./models/Specialty');
const Procedure = require('./models/Procedure');
require('dotenv').config();

// Specialty mapping based on procedure names and content
const specialtyMapping = {
  'endodontics': {
    keywords: ['root canal', 'apicoectomy', 'pulpotomy', 'vital pulp', 'hemisection', 'root amputation', 'root resection'],
    name: 'Endodontics',
    icon: 'Activity',
    color: 'bg-blue-50 border-blue-200'
  },
  'oral-surgery': {
    keywords: ['extraction', 'surgical', 'wisdom tooth', 'impacted', 'bone graft', 'sinus lift', 'implant', 'cysts', 'lesion removal', 'trauma', 'cancer surgery', 'orthognathic', 'tori removal', 'cleft'],
    name: 'Oral Surgery',
    icon: 'Scissors', 
    color: 'bg-red-50 border-red-200'
  },
  'prosthodontics': {
    keywords: ['crown', 'bridge', 'denture', 'veneer', 'inlay', 'onlay', 'full mouth rehabilitation', 'immediate dentures', 'reline'],
    name: 'Prosthodontics',
    icon: 'Crown',
    color: 'bg-purple-50 border-purple-200'
  },
  'periodontics': {
    keywords: ['scaling', 'root planing', 'periodontal', 'gum', 'gingivectomy', 'gingivoplasty', 'flap surgery', 'tissue graft', 'connective tissue', 'free gingival', 'crown lengthening', 'osseous', 'regenerative', 'guided tissue'],
    name: 'Periodontics',
    icon: 'Heart',
    color: 'bg-green-50 border-green-200'
  },
  'general-dentistry': {
    keywords: ['filling', 'amalgam', 'tooth colored', 'bonding', 'sealant', 'whitening', 'cleaning'],
    name: 'General Dentistry',
    icon: 'Shield',
    color: 'bg-orange-50 border-orange-200'
  },
  'orthodontics': {
    keywords: ['orthodontic', 'tad placement', 'surgical orthodontics', 'adjustment', 'appliance delivery'],
    name: 'Orthodontics',
    icon: 'Zap',
    color: 'bg-teal-50 border-teal-200'
  },
  'pedodontics': {
    keywords: ['pediatric', 'child', 'kids'],
    name: 'Pediatric Dentistry', 
    icon: 'Baby',
    color: 'bg-pink-50 border-pink-200'
  },
  'oral-medicine': {
    keywords: ['tmj', 'sleep apnea', 'nightguard', 'occlusal guard', 'oral cancer screening', 'biopsy'],
    name: 'Oral Medicine',
    icon: 'Stethoscope',
    color: 'bg-indigo-50 border-indigo-200'
  }
};

function categorizeByKeywords(procedureName, content) {
  const text = (procedureName + ' ' + content).toLowerCase();
  
  for (const [specialtyId, specialty] of Object.entries(specialtyMapping)) {
    for (const keyword of specialty.keywords) {
      if (text.includes(keyword.toLowerCase())) {
        return {
          id: specialtyId,
          name: specialty.name,
          icon: specialty.icon,
          color: specialty.color
        };
      }
    }
  }
  
  // Default to general dentistry if no match
  return {
    id: 'general-dentistry',
    name: 'General Dentistry',
    icon: 'Shield',
    color: 'bg-orange-50 border-orange-200'
  };
}

function extractStructuredData(text, filename) {
  // Clean up the text
  const cleanText = text.replace(/\s+/g, ' ').trim();
  
  // Extract procedure name from filename
  const procedureName = filename
    .replace('Post_Op_', '')
    .replace('.pdf', '')
    .replace(/_/g, ' ')
    .replace(/\s+\(\d+\)/g, '') // Remove (1), (2) etc.
    .trim();

  // Create unique ID
  const procedureId = procedureName.toLowerCase().replace(/[^a-z0-9]/g, '-').replace(/-+/g, '-');

  // Extract sections using various patterns
  const sections = {
    overview: '',
    immediateAftercare: [],
    dietRestrictions: [],
    warningSignsToCallDoctor: [],
    recoveryTimeline: [],
    medications: []
  };

  // Try to extract overview (first meaningful paragraph)
  const overviewMatch = cleanText.match(/(?:overview|description|procedure|what to expect)[:\s]+(.*?)(?=\n\n|\.|immediate|post|care|instructions)/i);
  if (overviewMatch) {
    sections.overview = overviewMatch[1].trim();
  } else {
    // Use first 200 characters as overview
    sections.overview = cleanText.substring(0, 200).trim() + '...';
  }

  // Extract immediate aftercare instructions
  const aftercarePatterns = [
    /(?:immediate|post-?op|after\s*care|instructions)[:\s]+(.*?)(?=diet|food|eating|warning|call|contact|medication)/is,
    /(?:immediately|first|today)[:\s]+(.*?)(?=diet|food|eating|warning|call|contact)/is
  ];
  
  for (const pattern of aftercarePatterns) {
    const match = cleanText.match(pattern);
    if (match) {
      const instructions = match[1].split(/[•\n\r]/).filter(item => item.trim().length > 10);
      sections.immediateAftercare = instructions.slice(0, 6).map(item => item.trim());
      break;
    }
  }

  // Extract diet restrictions
  const dietPatterns = [
    /(?:diet|food|eating|drink)[:\s]+(.*?)(?=warning|call|contact|medication|pain)/is,
    /(?:avoid|do not eat|don't eat)[:\s]+(.*?)(?=warning|call|contact)/is
  ];
  
  for (const pattern of dietPatterns) {
    const match = cleanText.match(pattern);
    if (match) {
      const restrictions = match[1].split(/[•\n\r]/).filter(item => item.trim().length > 10);
      sections.dietRestrictions = restrictions.slice(0, 5).map(item => item.trim());
      break;
    }
  }

  // Extract warning signs
  const warningPatterns = [
    /(?:warning|call|contact|emergency|concern)[:\s]+(.*?)(?=medication|pain|recovery|follow)/is,
    /(?:signs?|symptoms?|when to call)[:\s]+(.*?)(?=medication|recovery)/is
  ];
  
  for (const pattern of warningPatterns) {
    const match = cleanText.match(pattern);
    if (match) {
      const warnings = match[1].split(/[•\n\r]/).filter(item => item.trim().length > 10);
      sections.warningSignsToCallDoctor = warnings.slice(0, 5).map(item => item.trim());
      break;
    }
  }

  // Extract medications
  const medicationPatterns = [
    /(?:medication|medicine|drug|pain|ibuprofen|acetaminophen)[:\s]+(.*?)(?=recovery|follow|appointment)/is,
    /(?:take|prescribed)[:\s]+(.*?)(?=recovery|follow)/is
  ];
  
  for (const pattern of medicationPatterns) {
    const match = cleanText.match(pattern);
    if (match) {
      const meds = match[1].split(/[•\n\r]/).filter(item => item.trim().length > 10);
      sections.medications = meds.slice(0, 4).map(item => item.trim());
      break;
    }
  }

  // Create default content if extraction failed
  if (sections.immediateAftercare.length === 0) {
    sections.immediateAftercare = [
      'Follow your dentist\'s specific post-operative instructions',
      'Take prescribed medications as directed',
      'Apply ice pack to reduce swelling if recommended',
      'Maintain gentle oral hygiene as instructed'
    ];
  }

  if (sections.dietRestrictions.length === 0) {
    sections.dietRestrictions = [
      'Avoid hard or crunchy foods for 24-48 hours',
      'Eat soft foods and liquids initially', 
      'Avoid extremely hot or cold foods/drinks',
      'No smoking or alcohol during healing period'
    ];
  }

  if (sections.warningSignsToCallDoctor.length === 0) {
    sections.warningSignsToCallDoctor = [
      'Severe pain that worsens after 2-3 days',
      'Excessive bleeding that won\'t stop',
      'Signs of infection (fever, pus, bad taste)',
      'Unusual swelling that increases after 48 hours',
      'Persistent numbness beyond expected timeframe'
    ];
  }

  if (sections.medications.length === 0) {
    sections.medications = [
      'Take prescribed pain medication as directed',
      'Use anti-inflammatory medication to reduce swelling',
      'Complete antibiotic course if prescribed'
    ];
  }

  // Create recovery timeline based on procedure type
  const timelineMap = {
    'extraction': [
      { day: '1', activity: 'Blood clot formation - avoid disturbing' },
      { day: '2-3', activity: 'Peak swelling, gradual pain reduction' },
      { day: '7', activity: 'Soft tissue healing progresses' },
      { day: '14-21', activity: 'Complete initial healing' }
    ],
    'surgical': [
      { day: '1-3', activity: 'Initial healing and swelling reduction' },
      { day: '7-10', activity: 'Suture removal if placed' },
      { day: '2-4 weeks', activity: 'Soft tissue healing' },
      { day: '6-8 weeks', activity: 'Complete healing expected' }
    ],
    'implant': [
      { day: '1-7', activity: 'Initial healing phase' },
      { day: '2-6 weeks', activity: 'Soft tissue integration' },
      { day: '3-6 months', activity: 'Osseointegration process' },
      { day: '6+ months', activity: 'Ready for final restoration' }
    ],
    'default': [
      { day: '1-2', activity: 'Initial healing and adjustment' },
      { day: '3-7', activity: 'Gradual improvement in comfort' },
      { day: '1-2 weeks', activity: 'Most symptoms should resolve' },
      { day: '2+ weeks', activity: 'Complete healing expected' }
    ]
  };

  let timelineType = 'default';
  if (procedureName.toLowerCase().includes('extraction')) timelineType = 'extraction';
  else if (procedureName.toLowerCase().includes('surgical') || procedureName.toLowerCase().includes('implant')) timelineType = 'surgical';
  else if (procedureName.toLowerCase().includes('implant')) timelineType = 'implant';

  sections.recoveryTimeline = timelineMap[timelineType];

  // Estimate duration based on procedure type
  let duration = '3-7 days recovery';
  if (procedureName.toLowerCase().includes('implant')) duration = '3-6 months healing';
  else if (procedureName.toLowerCase().includes('surgical')) duration = '7-14 days recovery';
  else if (procedureName.toLowerCase().includes('extraction')) duration = '5-10 days recovery';
  else if (procedureName.toLowerCase().includes('root canal')) duration = '7-14 days recovery';

  return {
    id: procedureId,
    name: procedureName,
    duration: duration,
    ...sections
  };
}

async function processPDFs() {
  try {
    // Connect to MongoDB
    await mongoose.connect(process.env.MONGO_URL);
    console.log('Connected to MongoDB');

    const pdfDir = '/app/pdf-source';
    const pdfFiles = fs.readdirSync(pdfDir).filter(file => file.endsWith('.pdf'));
    
    console.log(`Found ${pdfFiles.length} PDF files to process`);

    // Process each PDF
    for (let i = 0; i < pdfFiles.length; i++) {
      const filename = pdfFiles[i];
      const filePath = path.join(pdfDir, filename);
      
      console.log(`Processing ${i + 1}/${pdfFiles.length}: ${filename}`);
      
      try {
        // Read and parse PDF
        const dataBuffer = fs.readFileSync(filePath);
        const data = await pdf(dataBuffer);
        const text = data.text;

        // Extract structured data
        const procedureData = extractStructuredData(text, filename);
        
        // Categorize into specialty
        const specialty = categorizeByKeywords(procedureData.name, text);
        
        // Add specialty information
        procedureData.specialty = specialty.id;
        procedureData.specialtyName = specialty.name;

        // Ensure specialty exists in database
        await Specialty.findOneAndUpdate(
          { id: specialty.id },
          {
            id: specialty.id,
            name: specialty.name,
            description: `${specialty.name} procedures and treatments`,
            icon: specialty.icon,
            color: specialty.color
          },
          { upsert: true, new: true }
        );

        // Save procedure to database
        await Procedure.findOneAndUpdate(
          { id: procedureData.id },
          procedureData,
          { upsert: true, new: true }
        );

        console.log(`✅ Processed: ${procedureData.name} -> ${specialty.name}`);

      } catch (error) {
        console.error(`❌ Error processing ${filename}:`, error.message);
      }
    }

    console.log('\n🎉 PDF processing complete!');
    
    // Get final counts
    const specialtyCount = await Specialty.countDocuments();
    const procedureCount = await Procedure.countDocuments();
    
    console.log(`📊 Final Database Stats:`);
    console.log(`   Specialties: ${specialtyCount}`);
    console.log(`   Procedures: ${procedureCount}`);

    process.exit(0);

  } catch (error) {
    console.error('❌ Error processing PDFs:', error);
    process.exit(1);
  }
}

processPDFs();

File: process_pdfs.py
#!/usr/bin/env python3

import os
import re
import subprocess
from pymongo import MongoClient
from datetime import datetime
import uuid

# Database connection
client = MongoClient('mongodb://localhost:27017')
db = client['test_database']

def extract_text_from_pdf(pdf_path):
    """Extract text from PDF using pdftotext"""
    try:
        result = subprocess.run(['pdftotext', pdf_path, '-'], 
                              capture_output=True, text=True, check=True)
        return result.stdout
    except subprocess.CalledProcessError as e:
        print(f"Error extracting text from {pdf_path}: {e}")
        return None

def parse_procedure_name(filename):
    """Extract procedure name from filename"""
    # Remove .pdf extension and Post_Op_ prefix
    name = filename.replace('.pdf', '').replace('Post_Op_', '').replace('Post-Operative Instructions_ ', '')
    # Replace underscores with spaces and title case
    name = name.replace('_', ' ').title()
    return name

def parse_pdf_content(text):
    """Parse PDF text content into structured data"""
    lines = [line.strip() for line in text.split('\n') if line.strip()]
    
    procedure_data = {
        'overview': '',
        'immediateAftercare': [],
        'dietRestrictions': [],
        'warningSignsToCallDoctor': [],
        'recoveryTimeline': [],
        'medications': []
    }
    
    current_section = 'overview'
    
    # First line is usually the procedure name/purpose
    if lines:
        procedure_data['name'] = lines[0]
        if len(lines) > 1 and lines[1].startswith('Purpose:'):
            procedure_data['overview'] = lines[1].replace('Purpose:', '').strip()
    
    # Parse sections
    for line in lines[1:]:
        line_lower = line.lower()
        
        # Identify sections
        if any(keyword in line_lower for keyword in ['first 24 hours', 'immediate', 'initially', 'day 1']):
            current_section = 'immediateAftercare'
            continue
        elif any(keyword in line_lower for keyword in ['diet', 'eating', 'food', 'what to eat', 'avoid eating']):
            current_section = 'dietRestrictions'
            continue
        elif any(keyword in line_lower for keyword in ['pain', 'swelling', 'bleeding', 'soreness', 'discomfort']):
            current_section = 'immediateAftercare'
            continue
        elif any(keyword in line_lower for keyword in ['oral hygiene', 'brushing', 'cleaning', 'rinse', 'floss']):
            current_section = 'immediateAftercare'
            continue
        elif any(keyword in line_lower for keyword in ['activity', 'exercise', 'rest', 'work']):
            current_section = 'immediateAftercare'
            continue
        elif any(keyword in line_lower for keyword in ['follow-up', 'follow up', 'contact', 'call', 'emergency', 'when to call']):
            current_section = 'warningSignsToCallDoctor'
            continue
        elif any(keyword in line_lower for keyword in ['medication', 'medicine', 'prescription', 'pain relief', 'antibiotic']):
            current_section = 'medications'
            continue
        
        # Add content to appropriate section
        if line.startswith('- ') or line.startswith('• '):
            # Bullet point
            content = line[2:].strip()
            if current_section == 'immediateAftercare':
                procedure_data['immediateAftercare'].append(content)
            elif current_section == 'dietRestrictions':
                procedure_data['dietRestrictions'].append(content)
            elif current_section == 'warningSignsToCallDoctor':
                procedure_data['warningSignsToCallDoctor'].append(content)
            elif current_section == 'medications':
                procedure_data['medications'].append(content)
        elif line and not line.endswith(':'):
            # Regular content line
            if current_section == 'overview' and not procedure_data['overview']:
                procedure_data['overview'] = line
            elif current_section == 'immediateAftercare':
                procedure_data['immediateAftercare'].append(line)
            elif current_section == 'dietRestrictions':
                procedure_data['dietRestrictions'].append(line)
            elif current_section == 'warningSignsToCallDoctor':
                procedure_data['warningSignsToCallDoctor'].append(line)
            elif current_section == 'medications':
                procedure_data['medications'].append(line)
    
    return procedure_data

def create_specialty_mapping():
    """Create specialty mappings for procedures"""
    specialty_map = {
        'oral_surgery': [
            'extraction', 'implant', 'bone graft', 'sinus lift', 'apicoectomy', 
            'biopsy', 'surgical', 'cyst', 'lesion', 'alveoloplasty'
        ],
        'restorative': [
            'filling', 'crown', 'bridge', 'onlay', 'inlay', 'amalgam', 'composite'
        ],
        'periodontics': [
            'gum', 'scaling', 'root planing', 'periodontal', 'gingivectomy', 
            'crown lengthening', 'flap'
        ],
        'endodontics': [
            'root canal', 'endodontic', 'pulpotomy', 'apicoectomy'
        ],
        'prosthodontics': [
            'denture', 'partial', 'full denture', 'overdenture', 'liner'
        ],
        'orthodontics': [
            'braces', 'orthodontic', 'retainer', 'spacer'
        ],
        'pedodontics': [
            'pediatric', 'child', 'children', 'pulpotomy', 'space maintainer'
        ],
        'cosmetic': [
            'whitening', 'bleaching', 'veneer', 'bonding'
        ]
    }
    
    return specialty_map

def determine_specialty(procedure_name):
    """Determine specialty based on procedure name"""
    specialty_map = create_specialty_mapping()
    procedure_lower = procedure_name.lower()
    
    for specialty, keywords in specialty_map.items():
        if any(keyword in procedure_lower for keyword in keywords):
            return specialty
    
    return 'general_dentistry'  # Default

def process_all_pdfs():
    """Process all PDFs and update database"""
    pdf_files = [f for f in os.listdir('/app') if f.endswith('.pdf') and f.startswith('Post_Op')]
    
    print(f"Found {len(pdf_files)} PDF files to process")
    
    # Clear existing procedures
    db.procedures.delete_many({})
    print("Cleared existing procedures from database")
    
    procedures_added = 0
    
    for pdf_file in sorted(pdf_files):  # Alphabetical order
        print(f"Processing: {pdf_file}")
        
        # Extract text
        pdf_path = f"/app/{pdf_file}"
        text_content = extract_text_from_pdf(pdf_path)
        
        if not text_content:
            print(f"Failed to extract text from {pdf_file}")
            continue
        
        # Parse content
        procedure_data = parse_pdf_content(text_content)
        procedure_name = parse_procedure_name(pdf_file)
        specialty = determine_specialty(procedure_name)
        
        # Create procedure document
        procedure_doc = {
            'id': str(uuid.uuid4()),
            'name': procedure_name,
            'specialty': specialty,
            'description': procedure_data['overview'] or f"Post-operative care instructions for {procedure_name.lower()}",
            'overview': procedure_data['overview'] or text_content[:500] + "...",
            'immediateAftercare': procedure_data['immediateAftercare'],
            'dietRestrictions': procedure_data['dietRestrictions'],
            'warningSignsToCallDoctor': procedure_data['warningSignsToCallDoctor'],
            'recoveryTimeline': procedure_data['recoveryTimeline'],
            'medications': procedure_data['medications'],
            'estimatedRecoveryTime': '1-2 weeks',  # Default
            'difficulty': 'moderate',  # Default
            'createdAt': datetime.utcnow(),
            'updatedAt': datetime.utcnow()
        }
        
        # Insert into database
        try:
            db.procedures.insert_one(procedure_doc)
            procedures_added += 1
            print(f"✅ Added: {procedure_name} ({specialty})")
        except Exception as e:
            print(f"❌ Failed to add {procedure_name}: {e}")
    
    print(f"\n🎉 Successfully processed {procedures_added} procedures!")
    
    # Update specialties collection
    update_specialties()

def update_specialties():
    """Update specialties collection with categories"""
    specialties = [
        {
            'id': 'oral_surgery',
            'name': 'Oral Surgery',
            'description': 'Surgical procedures including extractions, implants, and bone grafts',
            'icon': '🦷',
            'color': 'bg-red-50 border-red-200'
        },
        {
            'id': 'restorative',
            'name': 'Restorative Dentistry',
            'description': 'Fillings, crowns, bridges, and other restorative procedures',
            'icon': '🔧',
            'color': 'bg-blue-50 border-blue-200'
        },
        {
            'id': 'periodontics',
            'name': 'Periodontics',
            'description': 'Gum disease treatment and periodontal procedures',
            'icon': '🌿',
            'color': 'bg-green-50 border-green-200'
        },
        {
            'id': 'endodontics',
            'name': 'Endodontics',
            'description': 'Root canal therapy and endodontic treatments',
            'icon': '🎯',
            'color': 'bg-purple-50 border-purple-200'
        },
        {
            'id': 'prosthodontics',
            'name': 'Prosthodontics',
            'description': 'Dentures, partials, and prosthetic replacements',
            'icon': '🦾',
            'color': 'bg-orange-50 border-orange-200'
        },
        {
            'id': 'orthodontics',
            'name': 'Orthodontics',
            'description': 'Braces, aligners, and orthodontic treatments',
            'icon': '📐',
            'color': 'bg-pink-50 border-pink-200'
        },
        {
            'id': 'pedodontics',
            'name': 'Pediatric Dentistry',
            'description': 'Dental care specifically for children',
            'icon': '👶',
            'color': 'bg-yellow-50 border-yellow-200'
        },
        {
            'id': 'cosmetic',
            'name': 'Cosmetic Dentistry',
            'description': 'Whitening, veneers, and aesthetic procedures',
            'icon': '✨',
            'color': 'bg-indigo-50 border-indigo-200'
        },
        {
            'id': 'general_dentistry',
            'name': 'General Dentistry',
            'description': 'Routine cleanings and general dental procedures',
            'icon': '🏥',
            'color': 'bg-gray-50 border-gray-200'
        }
    ]
    
    # Clear and update specialties
    db.specialties.delete_many({})
    for specialty in specialties:
        db.specialties.insert_one(specialty)
    
    print("✅ Updated specialties collection")

if __name__ == "__main__":
    process_all_pdfs()